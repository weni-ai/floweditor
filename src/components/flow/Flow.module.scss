@import 'variables.module.scss';

.new_updates {

  > div {
    padding: $unnnic-spacing-stack-lg $unnnic-spacing-stack-md !important;
  }

  .content {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 20px;

    .title {
      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-black;
      line-height: $unnnic-line-height-md + $unnnic-font-size-title-sm;
    }

    .description {
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }


    .news_list {
      display: flex;
      gap: $unnnic-spacing-xs;
      flex-direction: column;
      align-items: center;
      text-align: center;
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }

    .footer {
      color: $unnnic-color-neutral-black;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-lg;
    }
  }

  .buttons {
    margin-top: $unnnic-spacing-stack-lg;
    display: flex;
    gap: $unnnic-spacing-stack-md;
    flex: 1;

    > * {
      flex: 1;

      button {
        width: 100%;
      }
    }
    
    .primary {
      background-color: $unnnic-color-brand-weni-soft;
    }

  }
}