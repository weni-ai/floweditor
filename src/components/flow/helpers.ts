import { Type, Types } from 'config/interfaces';
import { getType, getTypeConfig } from 'config/typeConfigs';
import { NodeEditorSettings } from 'store/nodeEditor';
import { Asset } from 'store/flowContext';
import { FlowIssue } from 'flowTypes';

export const determineTypeConfig = (nodeSettings: NodeEditorSettings): Type => {
  const renderNode = nodeSettings.originalNode;
  const node = renderNode && renderNode.node;

  if (nodeSettings.originalAction && nodeSettings.originalAction.type) {
    return getTypeConfig(nodeSettings.originalAction.type);
  } else if (node && node.actions && node.actions.length > 0) {
    return getTypeConfig(node.actions[node.actions.length - 1].type);
  } else {
    try {
      const type = getType(renderNode);
      const config = getTypeConfig(type);
      if (config.type !== Types.missing) {
        return config;
      }
      // tslint:disable-next-line:no-empty
    } catch (Error) {}
  }

  // Account for ghost nodes
  if (node && node.router) {
    return getTypeConfig(node.router.type);
  }

  throw new Error(`Couldn't determine type config for: ${node.uuid}`);
};

export const hasIssues = (
  issues: FlowIssue[],
  translating: boolean,
  language: Asset,
): boolean => {
  if (!issues || issues.length === 0) {
    return false;
  }

  if (!translating) {
    return !!issues.find((issue: FlowIssue) => !issue.language);
  } else {
    return !!issues.find((issue: FlowIssue) => issue.language === language.id);
  }
};
