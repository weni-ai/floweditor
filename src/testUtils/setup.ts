/* eslint-disable no-redeclare */
import { Console } from 'console';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

// Declare custom matcher types
declare global {
  namespace jest {
    interface Matchers<R> {
      toPointTo(renderNode: RenderNode): R;
      toHaveExitThatPointsTo(renderNode: RenderNode): R;
      toHaveInboundFrom(exit: Exit): R;
      toHaveExitWithDestination(): R;
      toHaveInboundConnections(): R;
      toHavePayload(action: string, payload: any): R;
      toHaveReduxActions(actions: string[]): R;
      toMatchCallSnapshot(snapshotName?: string): R;
    }
  }
}

// Ensure console logs are visible while running tests https://github.com/facebook/jest/issues/3853
global.console = new Console(process.stderr, process.stderr);

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
// tslint:disable-next-line:ban-types
(global as any).requestAnimationFrame = (callback: Function) => {
  setTimeout(callback, 0);
};
