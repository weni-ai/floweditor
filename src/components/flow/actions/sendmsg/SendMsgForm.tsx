/* eslint-disable @typescript-eslint/explicit-member-accessibility */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { react as bindCallbacks } from 'auto-bind';
import { AxiosResponse } from 'axios';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import {
  initializeForm as stateToForm,
  stateToAction,
  TOPIC_OPTIONS,
  INSTAGRAM_RESPONSE_TYPES,
  TAG_OPTIONS,
} from 'components/flow/actions/sendmsg/helpers';
import { ActionFormProps } from 'components/flow/props';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import { hasUseableTranslation } from 'components/form/assetselector/helpers';
import SwitchElement, {
  SwitchSizes,
} from 'components/form/switch/SwitchElement';
import MultiChoiceInput from 'components/form/multichoice/MultiChoice';
import SelectElement, {
  SelectOption,
} from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import { fetchAsset } from 'external';
import { Template, TemplateTranslation } from 'flowTypes';
import mutate from 'immutability-helper';
import * as React from 'react';
import { Asset } from 'store/flowContext';
import {
  FormState,
  mergeForm,
  StringArrayEntry,
  StringEntry,
  SelectOptionEntry,
  FormEntry,
} from 'store/nodeEditor';
import {
  MaxOfTenItems,
  Required,
  shouldRequireIf,
  validate,
} from 'store/validators';
import { range } from 'utils';

import styles from './SendMsgForm.module.scss';
import { hasFeature } from 'config/typeConfigs';
import { FeatureFilter } from 'config/interfaces';

import i18n from 'config/i18n';
import { Trans } from 'react-i18next';
import { Attachment, renderAttachments } from './attachments';

import { applyPureVueInReact } from 'veaury';
// @ts-ignore
import Unnnic from '@weni/unnnic-system';

const UnnnicIcon = applyPureVueInReact(Unnnic.unnnicIcon);

export interface SendMsgFormState extends FormState {
  message: StringEntry;
  quickReplies: StringArrayEntry;
  quickReplyEntry: StringEntry;
  sendAll: boolean;
  attachments: Attachment[];
  template: FormEntry;
  topic: SelectOptionEntry;
  templateVariables: StringEntry[];
  templateTranslation?: TemplateTranslation;
  instagramResponseType: SelectOptionEntry;
  postId: StringEntry;
  commentId: StringEntry;
  tagSelection: SelectOptionEntry;
}

export default class SendMsgForm extends React.Component<
  ActionFormProps,
  SendMsgFormState
> {
  private filePicker: any;

  constructor(props: ActionFormProps) {
    super(props);
    this.state = stateToForm(this.props.nodeSettings, this.props.assetStore);

    bindCallbacks(this, {
      include: [/^handle/, /^on/],
    });

    // initialize our templates if we have them
    if (this.state.template.value !== null) {
      fetchAsset(
        this.props.assetStore.templates,
        this.state.template.value.uuid,
      ).then((asset: Asset) => {
        if (asset !== null) {
          this.handleTemplateChanged([
            { ...this.state.template.value, ...asset.content },
          ]);
        }
      });
    }
  }

  public static contextTypes = {
    config: fakePropType,
  };

  private handleUpdate(
    keys: {
      text?: string;
      sendAll?: boolean;
      quickReplies?: string[];
    },
    submitting = false,
  ): boolean {
    const updates: Partial<SendMsgFormState> = {};
    if (keys.hasOwnProperty('text')) {
      updates.message = validate(
        i18n.t('forms.message', 'Message'),
        keys.text,
        [shouldRequireIf(submitting)],
      );
    }

    if (keys.hasOwnProperty('sendAll')) {
      updates.sendAll = keys.sendAll;
    }

    if (keys.hasOwnProperty('quickReplies')) {
      updates.quickReplies = validate(
        i18n.t('forms.quick_replies', 'Quick Replies'),
        keys.quickReplies,
        [MaxOfTenItems],
      );
    }

    const updated = mergeForm(this.state, updates) as SendMsgFormState;

    this.setState(updated);
    return updated.valid;
  }

  public handleMessageInput(event: React.KeyboardEvent) {
    return this.handleUpdate({ text: (event.target as any).value }, false);
  }

  public handleMessageUpdate(
    message: string,
    name: string,
    submitting = false,
  ): boolean {
    return this.handleUpdate({ text: message }, submitting);
  }

  public handleQuickRepliesUpdate(quickReplies: string[]): boolean {
    return this.handleUpdate({ quickReplies });
  }

  public handleSendAllUpdate(sendAll: boolean): boolean {
    return this.handleUpdate({ sendAll });
  }

  private handleSave(): void {
    // don't continue if our message already has errors
    if (hasErrors(this.state.message)) {
      return;
    }

    // make sure we validate untouched text fields and contact fields
    let valid = this.handleMessageUpdate(this.state.message.value, null, true);

    let templateVariables = this.state.templateVariables;
    // make sure we don't have untouched template variables
    this.state.templateVariables.forEach(
      (variable: StringEntry, num: number) => {
        const updated = validate(`Variable ${num + 1}`, variable.value, [
          Required,
        ]);
        templateVariables = mutate(templateVariables, {
          [num]: { $merge: updated },
        }) as StringEntry[];
        valid = valid && !hasErrors(updated);
      },
    );

    valid = valid && !hasErrors(this.state.quickReplyEntry);

    // Validate Instagram-specific fields based on response type
    if (this.state.instagramResponseType.value) {
      const responseType = this.state.instagramResponseType.value.value;

      // Map response types to their validation configuration
      const validationConfig: {
        [key: string]: {
          field: string;
          label: string;
          stateKey: keyof SendMsgFormState;
          value: any;
        };
      } = {
        tag_support: {
          field: 'tag_selection',
          label: i18n.t('forms.tags', 'Tags'),
          stateKey: 'tagSelection',
          value: this.state.tagSelection.value,
        },
        post_response: {
          field: 'post_id',
          label: i18n.t('forms.post_id', 'Post ID'),
          stateKey: 'postId',
          value: this.state.postId.value,
        },
        comment_reply: {
          field: 'comment_id',
          label: i18n.t('forms.comment_id', 'Comment ID'),
          stateKey: 'commentId',
          value: this.state.commentId.value,
        },
      };

      // Only validate if we have a config for this response type
      if (responseType !== 'none' && validationConfig[responseType]) {
        const config = validationConfig[responseType];
        const validatedEntry = validate(config.label, config.value, [Required]);

        // Update state for the validated field
        const stateUpdate = {} as any;
        stateUpdate[config.stateKey] = validatedEntry;
        this.setState(stateUpdate);

        // Check if validation passed
        valid = valid && !hasErrors(validatedEntry);
      }
    }

    if (valid) {
      this.props.updateAction(
        stateToAction(this.props.nodeSettings, this.state),
      );
      // notify our modal we are done
      this.props.onClose(false);
    } else {
      this.setState({ templateVariables, valid });
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.confirm'), onClick: this.handleSave },
      tertiary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true),
      },
    };
  }

  private handleTemplateChanged(selected: any[]): void {
    const template = selected ? selected[0] : null;

    if (!template) {
      this.setState({
        template: { value: null },
        templateTranslation: null,
        templateVariables: [],
      });
    } else {
      const templateTranslation = template.translations[0];

      const templateVariables =
        this.state.templateVariables.length === 0 ||
        (this.state.template.value &&
          this.state.template.value.uuid !== template.uuid)
          ? range(0, templateTranslation.variable_count).map(() => {
              return {
                value: '',
              };
            })
          : this.state.templateVariables;

      this.setState({
        template: { value: template },
        templateTranslation,
        templateVariables,
      });
    }
  }

  private handleTemplateVariableChanged(
    updatedText: string,
    num: number,
  ): void {
    const entry = validate(`Variable ${num + 1}`, updatedText, [Required]);
    const templateVariables = mutate(this.state.templateVariables, {
      $merge: { [num]: entry },
    }) as StringEntry[];
    this.setState({ templateVariables });
  }

  private handleShouldExcludeTemplate(template: any): boolean {
    return !hasUseableTranslation(template as Template);
  }

  private renderTopicConfig(): JSX.Element {
    return (
      <>
        <p>
          {i18n.t(
            'forms.send_msg_facebook_warning',
            'Sending bulk messages over a Facebook channel requires that a topic be specified if the user has not sent a message in the last 24 hours. Setting a topic to use over Facebook is especially important for the first message in your flow.',
          )}
        </p>
        <SelectElement
          key={'fb_method_select'}
          name={i18n.t('forms.method', 'Method')}
          entry={this.state.topic}
          onChange={this.handleTopicUpdate}
          options={TOPIC_OPTIONS}
          placeholder={i18n.t(
            'forms.send_msg_facebook_topic_placeholder',
            'Select a topic to use over Facebook',
          )}
          clearable={true}
        />
      </>
    );
  }

  private handleTopicUpdate(topic: SelectOption) {
    this.setState({ topic: { value: topic } });
  }

  private handleInstagramResponseTypeChanged(selected: SelectOption): void {
    // If "None" is selected, clear all Instagram-related values
    if (selected && selected.value === 'none') {
      this.setState({
        instagramResponseType: { value: null },
        postId: { value: '' },
        commentId: { value: '' },
        tagSelection: { value: null },
      });
      return;
    }

    this.setState({
      instagramResponseType: { value: selected },
    });
  }

  private handlePostIdChanged(value: string): void {
    const entry = validate(i18n.t('forms.post_id', 'Post ID'), value, [
      Required,
    ]);
    this.setState({ postId: entry });
  }

  private handleCommentIdChanged(value: string): void {
    const entry = validate(i18n.t('forms.comment_id', 'Comment ID'), value, [
      Required,
    ]);
    this.setState({ commentId: entry });
  }

  private handleTagSelectionChanged(selected: SelectOption): void {
    const entry = validate(i18n.t('forms.tags', 'Tags'), selected, [Required]);

    this.setState({
      tagSelection: entry,
    });
  }

  private renderResponseTypeContent(responseType: string): JSX.Element {
    // Define a type for the configuration
    type ConfigItem = {
      description: string;
      fieldLabel: string;
      inputType: string;
      placeholder?: string;
    };

    type ConfigMap = {
      [key: string]: ConfigItem;
    };

    // Common configuration for each response type
    const config: ConfigMap = {
      tag_support: {
        description: i18n.t(
          'forms.tag_support_description',
          'Allows categorizing direct messages (DMs) by adding predefined tags from Meta. This helps organize and segment contacts within Instagram.',
        ),
        fieldLabel: i18n.t('forms.select_tags', 'Select the tags'),
        inputType: 'select',
      },
      post_response: {
        description: i18n.t(
          'forms.post_response_description',
          'When a specific post is identified, this action enables sending a direct message (DM) to a user who interacted with it, referencing the original post.',
        ),
        fieldLabel: i18n.t('forms.post_id', 'Post ID'),
        inputType: 'text',
        placeholder: 'E.g., 17948912345678901',
      },
      comment_reply: {
        description: i18n.t(
          'forms.comment_reply_description',
          "When a comment is made on a post, this action allows replying directly within the post's comment section, keeping the interaction public.",
        ),
        fieldLabel: i18n.t('forms.comment_id', 'Comment ID'),
        inputType: 'text',
        placeholder: 'E.g., 17948912345678901',
      },
    };

    // If response type doesn't match any known type, return null
    if (!responseType || !config[responseType]) {
      return null;
    }

    const instagramTypeConfig = config[responseType];

    return (
      <>
        <div className={styles.disclaimer}>
          <UnnnicIcon icon="alert-circle-1-1" scheme="feedback-blue" />
          <p className={styles.disclaimer_text}>
            {instagramTypeConfig.description}
          </p>
        </div>

        {instagramTypeConfig.inputType === 'select' ? (
          <SelectElement
            name={i18n.t('forms.tags', 'Tags')}
            entry={this.state.tagSelection}
            onChange={this.handleTagSelectionChanged}
            options={TAG_OPTIONS}
            placeholder={i18n.t('forms.select_tag', 'Select a tag')}
            clearable={false}
            showLabel={true}
          />
        ) : (
          <TextInputElement
            name={instagramTypeConfig.fieldLabel}
            showLabel={true}
            onChange={
              responseType === 'post_response'
                ? this.handlePostIdChanged
                : this.handleCommentIdChanged
            }
            entry={
              responseType === 'post_response'
                ? this.state.postId
                : this.state.commentId
            }
            placeholder={instagramTypeConfig.placeholder}
          />
        )}
      </>
    );
  }

  private renderInstagramConfig(): JSX.Element {
    const responseType = this.state.instagramResponseType.value
      ? this.state.instagramResponseType.value.value
      : null;

    return (
      <>
        <SelectElement
          key="instagram_response_type"
          name={i18n.t('forms.response_type', 'Choose a response Type')}
          entry={this.state.instagramResponseType}
          onChange={this.handleInstagramResponseTypeChanged}
          options={INSTAGRAM_RESPONSE_TYPES}
          placeholder={i18n.t(
            'forms.select_response_type',
            'Select a response type',
          )}
          showLabel={true}
        />

        {responseType &&
          responseType !== 'none' &&
          this.renderResponseTypeContent(responseType)}
      </>
    );
  }

  private renderTemplateConfig(): JSX.Element {
    return (
      <>
        <p>
          {i18n.t(
            'forms.whatsapp_warning',
            'Sending messages over a WhatsApp channel requires that a template be used if you have not received a message from a contact in the last 24 hours. Setting a template to use over WhatsApp is especially important for the first message in your flow.',
          )}
        </p>
        <AssetSelector
          name={i18n.t('forms.template', 'template')}
          noOptionsMessage="No templates found"
          assets={this.props.assetStore.templates}
          entry={this.state.template}
          onChange={this.handleTemplateChanged}
          shouldExclude={this.handleShouldExcludeTemplate}
          searchable={true}
          formClearable={true}
        />
        {this.state.templateTranslation ? (
          <>
            <div className={styles.template_text}>
              {this.state.templateTranslation.content}
            </div>
            {range(0, this.state.templateTranslation.variable_count).map(
              (num: number) => {
                return (
                  <div className={styles.variable} key={'tr_arg_' + num}>
                    <TextInputElement
                      name={`${i18n.t('forms.variable', 'Variable')} ${num +
                        1}`}
                      showLabel={false}
                      placeholder={`${i18n.t(
                        'forms.variable',
                        'Variable',
                      )} ${num + 1}`}
                      onChange={(updatedText: string) => {
                        this.handleTemplateVariableChanged(updatedText, num);
                      }}
                      entry={this.state.templateVariables[num]}
                      autocomplete={true}
                    />
                  </div>
                );
              },
            )}
          </>
        ) : null}
      </>
    );
  }

  private handleAttachmentUploaded(response: AxiosResponse) {
    const attachments: any = mutate(this.state.attachments, {
      $push: [
        { type: response.data.type, url: response.data.url, uploaded: true },
      ],
    });
    this.setState({ attachments });
  }

  private handleAttachmentChanged(index: number, type: string, url: string) {
    let attachments: any = this.state.attachments;
    if (index === -1) {
      attachments = mutate(attachments, {
        $push: [{ type, url }],
      });
    } else {
      attachments = mutate(attachments, {
        [index]: {
          $set: { type, url },
        },
      });
    }

    this.setState({ attachments });
  }

  private handleAttachmentRemoved(index: number) {
    const attachments: any = mutate(this.state.attachments, {
      $splice: [[index, 1]],
    });
    this.setState({ attachments });
  }

  private instagramErrors(): boolean {
    if (!this.state.instagramResponseType.value) {
      return false;
    }

    const responseType = this.state.instagramResponseType.value.value;

    // Map response types to their state fields to check for errors
    const validationMap: { [key: string]: keyof SendMsgFormState } = {
      tag_support: 'tagSelection',
      post_response: 'postId',
      comment_reply: 'commentId',
    };

    if (responseType !== 'none' && validationMap[responseType]) {
      const fieldToCheck = validationMap[responseType];
      return hasErrors(this.state[fieldToCheck] as any);
    }

    return false;
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const quickReplies: Tab = {
      name: i18n.t('forms.quick_replies'),
      body: (
        <>
          <div
            className={`${styles.quick_reply_summary} u font secondary body-md color-neutral-cloudy`}
          >
            {i18n.t(
              'forms.quick_replies_summary',
              'Quick Replies are made into buttons for supported channels. For example, when asking a question, you might add a Quick Reply for "Yes" and one for "No".',
            )}
            <br />
            <br />
            {i18n.t(
              'forms.quick_replies_limit',
              'You can add up to 10 Quick Replies.',
            )}
          </div>

          <MultiChoiceInput
            name={i18n.t(
              'forms.add_quick_reply',
              'Add a new Quick Reply and press enter',
            )}
            helpText={<Trans i18nKey="forms.quick_replies" />}
            items={this.state.quickReplies}
            entry={this.state.quickReplyEntry}
            onChange={this.handleQuickRepliesUpdate}
          />
        </>
      ),
      checked: this.state.quickReplies.value.length > 0,
      hasErrors: hasErrors(this.state.quickReplyEntry),
    };

    const attachments: Tab = {
      name: i18n.t('forms.attachments', 'Attachments'),
      body: renderAttachments(
        this.context.config.endpoints.attachments,
        this.state.attachments,
        this.handleAttachmentUploaded,
        this.handleAttachmentChanged,
        this.handleAttachmentRemoved,
      ),
      checked: this.state.attachments.length > 0,
    };

    const advanced: Tab = {
      name: i18n.t('forms.advanced', 'Advanced'),
      body: (
        <>
          <div
            className={`u font secondary body-md color-neutral-cloudy ${styles.title}`}
          >
            {i18n.t(
              'forms.all_destinations_title',
              'Send a message to all destinations known for this contact.',
            )}
          </div>

          <SwitchElement
            name={i18n.t(
              'forms.all_destinations',
              'Send a message to all destinations',
            )}
            title={i18n.t(
              'forms.all_destinations',
              'Send a message to all destinations',
            )}
            checked={this.state.sendAll}
            description={i18n.t(
              'forms.all_destinations_description',
              "If you aren't sure what this means, leave it unchecked.",
            )}
            onChange={this.handleSendAllUpdate}
            size={SwitchSizes.small}
          />
        </>
      ),
      checked: this.state.sendAll,
    };

    const tabs = [quickReplies, attachments, advanced];

    if (hasFeature(this.context.config, FeatureFilter.HAS_INSTAGRAM)) {
      const templates: Tab = {
        name: 'Instagram',
        body: this.renderInstagramConfig(),
        checked: this.state.instagramResponseType.value !== null,
        hasErrors: this.instagramErrors(),
      };
      tabs.splice(0, 0, templates);
    }

    if (hasFeature(this.context.config, FeatureFilter.HAS_WHATSAPP)) {
      const templates: Tab = {
        name: 'WhatsApp',
        body: this.renderTemplateConfig(),
        checked: this.state.template.value != null,
        hasErrors: !!this.state.templateVariables.find((entry: StringEntry) =>
          hasErrors(entry),
        ),
      };
      tabs.splice(0, 0, templates);
    }

    if (hasFeature(this.context.config, FeatureFilter.HAS_FACEBOOK)) {
      const templates: Tab = {
        name: 'Facebook',
        body: this.renderTopicConfig(),
        checked: this.state.topic.value != null,
      };
      tabs.splice(0, 0, templates);
    }

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList
          __className=""
          initialType={typeConfig}
          onChange={this.props.onTypeChange}
          nodeSettings={this.props.nodeSettings}
        />
        <TextInputElement
          name={i18n.t('forms.message', 'Message')}
          showLabel={true}
          counter="sms"
          onChange={this.handleMessageUpdate}
          entry={this.state.message}
          autocomplete={true}
          focus={true}
          textarea={true}
          placeholder={i18n.t('forms.type_here', 'Type here...')}
        />
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
