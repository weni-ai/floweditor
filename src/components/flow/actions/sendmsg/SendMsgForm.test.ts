import SendMsgForm from 'components/flow/actions/sendmsg/SendMsgForm';
import { ActionFormProps } from 'components/flow/props';
import { composeComponentTestUtils, mock } from 'testUtils';
import {
  createSendMsgAction,
  getActionFormProps,
} from 'testUtils/assetCreators';
import * as utils from 'utils';
import { shallowToJson } from 'enzyme-to-json';

const { setup } = composeComponentTestUtils<ActionFormProps>(
  SendMsgForm,
  getActionFormProps(createSendMsgAction()),
);

mock(utils, 'createUUID', utils.seededUUIDs());

describe(SendMsgForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = setup(true);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true);

      instance.handleMessageUpdate('What is your favorite color?', []);
      instance.handleQuickRepliesUpdate(['red', 'green', 'blue']);
      instance.handleSendAllUpdate(true);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.updateAction).toHaveBeenCalled();
      expect(props.updateAction).toMatchSnapshot();
    });

    it('should allow switching from router', () => {
      const { instance, props } = setup(true, {
        $merge: { updateAction: vi.fn() },
        nodeSettings: { $merge: { originalAction: null } },
      });

      instance.handleMessageUpdate('What is your favorite color?', []);
      instance.handleSave();

      expect(props.updateAction).toMatchSnapshot();
    });
  });

  describe('cancel', () => {
    it('should cancel without changes', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateAction: vi.fn() },
      });
      instance.handleMessageUpdate("Don't save me bro", []);
      instance.getButtons().tertiary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateAction).not.toHaveBeenCalled();
    });
  });
});
