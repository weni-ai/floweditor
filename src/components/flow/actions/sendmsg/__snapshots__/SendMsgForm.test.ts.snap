// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SendMsgForm > render > should render 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "tertiary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="send_msg"
  tabs={
    [
      {
        "body": <React.Fragment>
          <p>
            Sending messages over a WhatsApp channel requires that a template be used if you have not received a message from a contact in the last 24 hours. Setting a template to use over WhatsApp is especially important for the first message in your flow.
          </p>
          <AssetSelector
            entry={
              {
                "value": null,
              }
            }
            formClearable={true}
            name="template"
            noOptionsMessage="No templates found"
            onChange={[Function]}
            searchable={true}
            shouldExclude={[Function]}
          />
        </React.Fragment>,
        "checked": false,
        "hasErrors": false,
        "name": "WhatsApp",
      },
      {
        "body": <React.Fragment>
          <div
            className="_quick_reply_summary_cc747c u font secondary body-md color-neutral-cloudy"
          >
            Quick Replies are made into buttons for supported channels. For example, when asking a question, you might add a Quick Reply for "Yes" and one for "No".
            <br />
            <br />
            You can add up to 10 Quick Replies.
          </div>
          <MultiChoiceInput
            entry={
              {
                "value": "",
              }
            }
            helpText={
              <Trans
                i18nKey="forms.quick_replies"
              />
            }
            items={
              {
                "value": [],
              }
            }
            name="Add a new Quick reply and press Enter"
            onChange={[Function]}
          />
        </React.Fragment>,
        "checked": false,
        "hasErrors": false,
        "name": "Quick Replies",
      },
      {
        "body": <React.Fragment>
          <div
            className="u font secondary body-md color-neutral-cloudy"
          >
            Add an attachment to each message. The attachment can be a file you upload or a dynamic URL using expressions and variables from your Flow.
          </div>
          <div
            className="_url_attachment_container_0ed93f"
          >
            <div
              className="_label_0ed93f u font secondary body-md color-neutral-cloudy"
            >
              File
               
              1
            </div>
            <div
              className="_url_attachment_0ed93f"
            >
              <div
                className="_type_choice_0ed93f"
              >
                <SelectElement
                  entry={
                    {
                      "value": null,
                    }
                  }
                  name="Type Options"
                  onChange={[Function]}
                  options={
                    [
                      {
                        "name": "Image URL",
                        "value": "image",
                      },
                      {
                        "name": "Audio URL",
                        "value": "audio",
                      },
                      {
                        "name": "Video URL",
                        "value": "video",
                      },
                      {
                        "name": "PDF Document URL",
                        "value": "application",
                      },
                      {
                        "name": "Upload Attachment",
                        "value": "upload",
                      },
                    ]
                  }
                  placeholder="Select the file type"
                  style="sm"
                />
              </div>
              <React.Fragment>
                <div
                  className="_url_0ed93f"
                />
                <div
                  className="_remove_0ed93f"
                />
              </React.Fragment>
            </div>
          </div>
          <input
            onChange={[Function]}
            style={
              {
                "display": "none",
              }
            }
            type="file"
          />
        </React.Fragment>,
        "checked": false,
        "name": "Attachments",
      },
      {
        "body": <React.Fragment>
          <div
            className="u font secondary body-md color-neutral-cloudy _title_cc747c"
          >
            Send a message to all destinations known for this contact.
          </div>
          <SwitchElement
            checked={false}
            description="If you aren't sure what this means, leave it unchecked."
            name="Send a message to all destinations"
            onChange={[Function]}
            size="small"
            title="Send a message to all destinations"
          />
        </React.Fragment>,
        "checked": false,
        "name": "Advanced",
      },
    ]
  }
  title="Send message"
>
  <TypeList
    __className=""
    initialType={
      {
        "component": [Function],
        "description": "Send the contact a message",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "text",
          "quick_replies",
          "templating.variables",
        ],
        "massageForDisplay": [Function],
        "name": "Send message",
        "type": "send_msg",
      }
    }
    nodeSettings={
      {
        "originalAction": {
          "all_urns": false,
          "attachments": [],
          "quick_replies": [],
          "templating": null,
          "text": "Hey!",
          "type": "send_msg",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [
              {
                "all_urns": false,
                "attachments": [],
                "quick_replies": [],
                "templating": null,
                "text": "Hey!",
                "type": "send_msg",
                "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
              },
            ],
            "exits": [
              {
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
            ],
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "ui": {
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "execute_actions",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <TextInputElement
    autocomplete={true}
    counter="sms"
    entry={
      {
        "value": "Hey!",
      }
    }
    focus={true}
    name="Message"
    onChange={[Function]}
    placeholder="Type here..."
    showLabel={true}
    textarea={true}
  />
</Dialog>
`;

exports[`SendMsgForm > updates > should allow switching from router 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "all_urns": false,
        "attachments": [],
        "quick_replies": [],
        "text": "What is your favorite color?",
        "type": "send_msg",
        "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`SendMsgForm > updates > should save changes 1`] = `
{
  "attachments": [],
  "commentId": {
    "value": "",
  },
  "instagramResponseType": {
    "value": null,
  },
  "message": {
    "validationFailures": [],
    "value": "What is your favorite color?",
  },
  "postId": {
    "value": "",
  },
  "quickReplies": {
    "validationFailures": [],
    "value": [
      "red",
      "green",
      "blue",
    ],
  },
  "quickReplyEntry": {
    "value": "",
  },
  "sendAll": true,
  "tagSelection": {
    "value": null,
  },
  "template": {
    "value": null,
  },
  "templateVariables": [],
  "topic": {
    "value": undefined,
  },
  "valid": true,
}
`;

exports[`SendMsgForm > updates > should save changes 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "all_urns": true,
        "attachments": [],
        "quick_replies": [
          "red",
          "green",
          "blue",
        ],
        "text": "What is your favorite color?",
        "type": "send_msg",
        "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
